eclass/dkms.eclass

```diff
@@ -0,0 +1,527 @@
+# @DESCRIPTION:
+# Uses linux-mod-r1's modlist and modargs to construct a DKMS
+# configuration file. By default DKMS adds the 'KERNELRELEASE='
+# variable to all make commands. Some Makefiles will behave
+# differently when this variable is set, if this leads to problems
+# pass the --no-kernelrelease argument to this function to suppress
+# the addition of 'KERNELRELEASE=' to the calls to make at runtime.
+# By default the created DKMS configuration file will enable
+# automatic installation of all kernel modules. To disable this add
+# the --no-autoinstall argument. Does nothing if USE=dkms is disabled.
+dkms_autoconf() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	use dkms || return 0
+	[[ -z ${MODULES_OPTIONAL_IUSE} ]] ||
+		use "${MODULES_OPTIONAL_IUSE#+}" || return 0
+
+	local arg autoinstall=1 make_command=make
+	[[ ${#} -gt 2 ]] && die "${FUNCNAME}: too many arguments"
+	for arg in "${@}"; do
+		case ${arg} in
+			--no-kernelrelease)
+				# Per DKMS manual, quoting disables setting KERNELRELEASE
+				make_command=\'make\'
+			;;
+			--no-autoinstall)
+				autoinstall=
+			;;
+			*)
+				die "${FUNCNAME}: invalid argument ${arg}"
+			;;
+		esac
+	done
+
+	modules_process_modlist
+
+	local edkms_args index mod name package
+	for mod in "${modlist[@]}"; do
+		name=${mod%%=*}
+		mod=${mod#"${name}"}
+		IFS=: read -ra mod <<<"${mod#=}"
+
+		pushd "${mod[1]}" >/dev/null || die
+		if [[ -f dkms.conf ]]; then
+			if grep -qE "^MAKE(\[0\]|)=" dkms.conf; then
+				dkms_gentoofy_conf dkms.conf
+			else
+				dkms_sanitize_makeargs
+				echo "MAKE[0]=\"${make_command} ${edkmsargs[*]} ${mod[3]}\"" \
+					>> dkms.conf || die
+			fi
+
+			# Find the index of an existing module, else find the
+			# first available index.
+			index=$(
+				source dkms.conf &>/dev/null ||
+					die "${FUNCNAME}: invalid dkms.conf at ${PWD}"
+				local i
+				for i in "${!BUILT_MODULE_NAME[@]}"; do
+					if [[ ${name} == ${BUILT_MODULE_NAME[${i}]} ]]
+					then
+						echo ${i} || die
+						exit 0
+					fi
+				done
+				echo ${#BUILT_MODULE_NAME[@]} || die
+			) || continue
+		else
+			# If the kernel modules are in a subdir add this to the
+			# DKMS package name identifier to ensure it is unique.
+			# There may be multiple subdirs with kernel modules.
+			if [[ ${PWD} == ${S} ]]; then
+				package=${PN}
+			else
+				package=${PN}_${name}
+			fi
+			dkms_sanitize_makeargs
+			cat <<-EOF > dkms.conf || die
+				PACKAGE_NAME=${package}
+				PACKAGE_VERSION=${PV}
+				MAKE[0]="${make_command} ${edkmsargs[*]} ${mod[3]}"
+			EOF
+			if [[ -n ${autoinstall} ]]; then
+				echo "AUTOINSTALL=yes" >> dkms.conf
+			else
+				echo "AUTOINSTALL=no" >> dkms.conf
+			fi
+			index=0
+		fi
+
+		# DKMS enforces that the install target starts with one of
+		# these options.
+		local dest=${mod[0]}
+		if ! [[ ${dest} == /kernel* || ${dest} == /updates* ||
+			${dest} == /extra* ]]
+		then
+			dest=/extra/${dest}
+		fi
+
+		cat <<-EOF >> dkms.conf || die
+
```

Comment:
The new line at the start looks off. If it's intentional, add a comment header instead?

Reply:
It's intentional, I encountered one instance of an upstream stub `dkms.conf` that was missing a newline at the end of the file which lead to a mess when trying to append to it. The extra new line prevents a mess in this case. I'll add a comment.